lib/amazonka/src/Amazonka/Auth/SSO.hs

```diff
@@ -0,0 +1,111 @@
+-- |
+-- Module      : Amazonka.Auth.SSO
+-- Copyright   : (c) 2013-2021 Brendan Hay
+-- License     : Mozilla Public License, v. 2.0.
+-- Maintainer  : Brendan Hay <brendan.g.hay+amazonka@gmail.com>
+-- Stability   : provisional
+-- Portability : non-portable (GHC extensions)
+module Amazonka.Auth.SSO where
+
+import Amazonka.Auth.Background (fetchAuthInBackground)
+import Amazonka.Auth.Exception
+import qualified Amazonka.Crypto as Crypto
+import Amazonka.Data.Sensitive
+import Amazonka.Data.Time (ISO8601, Time (..))
+import Amazonka.Env (Env, Env' (..))
+import Amazonka.HTTP (retryRequest)
+import Amazonka.Lens ((^.))
+import Amazonka.Prelude
+import Amazonka.SSO.GetRoleCredentials as SSO
+import qualified Amazonka.SSO.Types as SSO (RoleCredentials (..))
+import Amazonka.Types
+import qualified Control.Exception as Exception
+import Control.Exception.Lens (handling_, _IOException)
+import Control.Monad.Trans.Resource (runResourceT)
+import Data.Aeson (FromJSON, decodeFileStrict)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import Data.Time.Clock.POSIX (posixSecondsToUTCTime)
+import qualified Network.HTTP.Client as Client
+
+data Cache = Cache
+  { accessToken :: Text,
+    region :: Region
+  }
+  deriving stock (Generic)
+  deriving anyclass (FromJSON)
+
+fromSSO ::
+  (MonadIO m, Foldable withAuth) =>
+  FilePath ->
+  Region ->
+  Text ->
+  Text ->
+  Env' withAuth ->
+  m Env
+fromSSO cache ssoRegion accountId roleName env = do
+  auth <- liftIO $ fetchAuthInBackground getCredentials
+  pure $ env {envAuth = Identity auth}
+  where
+    getCredentials = do
+      Cache {..} <- readCache cache
+
+      let ssoEnv = env {envRegion = ssoRegion}
+
+      resp <-
+        runResourceT $
+          sendWithoutAuth ssoEnv $
+            SSO.newGetRoleCredentials roleName accountId accessToken
+
+      let mCreds = do
+            rc <- resp ^. SSO.getRoleCredentialsResponse_roleCredentials
+            roleCredentialsToAuthEnv rc
```

Comment:
Would you like to fix up the service definition, or shall we proceed as-is?

Reply:
It's up to you. I'd need some hand-holding, but I'm happy to do it if you'd like.

Reply:
OK, if you're willing to wait for this, then let's do it properly.

Reply:
#759 - I'll merge in a day or so in case anyone has comments, then regenerate.